src/include/threadpool/worker_pool.h

```diff
@@ -34,24 +41,33 @@ class WorkerPool {
         should_shutdown_(false),
         task_queue_(task_queue) {}
 
-  void Startup() {
-    for (size_t i = 0; i < num_workers_; i++) {
-      workers_.emplace_back(WorkerFunc, &should_shutdown_, task_queue_);
-    }
-  }
+  /**
+   * Start the threads to begin polling the task queue
+   */
+  void Startup();
 
-  void Shutdown() {
-    should_shutdown_ = true;
-    for (auto &worker : workers_) {
-      worker.join();
-    }
-    workers_.clear();
+  /**
+   * Tell the threads to shutdown after the task queue is empty
+   */
+  void Shutdown();
+
+  /**
+   * Returns true if this pool has been marked for shutdown
+   * @return
+   */
+  bool ShouldShutdown() const {
+    return (should_shutdown_);
   }
 
  private:
   std::vector<std::thread> workers_;
   size_t num_workers_;
-  std::atomic_bool should_shutdown_;
+
+  // This used to be a std::atomic_bool
+  // It doesn't need to be because don't care about updating
+  // this flag atomically.
+  bool should_shutdown_;
```

Comment:
I'm not sure this is safe. Making it a vanilla member variable allows the compiler to optimize away the loop check in `WorkerFunc`. While we don't need concurrent updates to the flag, we do need every read to go through cache/memory. Atomics provide this (and `volatile` I think, but it's probably safer to use C++ atomics).

Reply:
It doesn't matter if it misses the update to the flag. It will loop back and check it again. Every read does not need to be consistent:

https://github.com/cmu-db/peloton/pull/1161/files/f57ef48227ef3eadadc7a518aef4bbda35c4c6d7#diff-01c2e44c857eb9c764c6c1fe0f244677R23